[stable9] Setupfs before access a users keys

DeepDiver1975 opened the pull request:
alternative approach for #26818 

@PVince81 @felixboehm 


## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.


…he keys

PVince81 commented:
original issue confirmed with your hack and also LDAP + homeDirectory on stable9

PVince81 commented:
This PR works for newly created shared files. Old files don't have the correct keys and can't be repaired without recreating or resharing (there is no other way).

I think `setupFS` and `tearDown` here is risky, because it removes all mount points and subsequent code might expect these to still exist. Let's try with `initMountPoints` instead.

DeepDiver1975 commented on lib/private/encryption/keys/storage.php:
```diff
@@ -72,14 +72,12 @@ public function __construct(View $view, Util $util) {
 		if (!empty($uid) && $uid !== $currentUser) {
-			\OC_Util::tearDownFS();
-			\OC_Util::setupFS($currentUser);
+			\OC\Files\Filesystem::initMountPoints($uid);
```
$currentUser

PVince81 replied:
Hmm, really? The FS of the current user is usually already setup.
What we need is setup the FS for the target user `$uid`. Strangely this fix works for me.

PVince81 replied:
To put it into perspective: let's say "user1" is logged in, so usually `setupFS("user1")` was already called early on. But "user1" shares a file with "user2" (or creates a file in a shared folder), so we also need to mount the home folder of "user2" to have proper access to the keys and files.

I suspect that the mounting already happens later in the sharing code, but if this enc code runs earlier than the sharing code then we need to do it too.

DeepDiver1975 replied:
wrong turn in my brain
I guess we don't need the second block - this was used for teardowns

PVince81 replied:
I'll also replace `empty($uid)`, remember that `empty("0")` evaluates to true...

PVince81 commented:
@DeepDiver1975 I redid a test run with folder shares, file shares and also creating a file inside an already shared folder and this fix as per my modification does work. This is with LDAP and an alternative home folder.

PVince81 commented:
Also retested with the hash folder hack, my fix makes it work too.

Commit:
The file system of a user has to be properly setup before accessing the keys

PVince81 commented:
Fixed and retested with the md5 hack variant, still works.

Squash ?

DeepDiver1975 commented:
Yes. Squash, merge at will and port it all around :party:

DeepDiver1975 commented:
@PVince81 squashed - merge now or later? Your call.
Will prepare ports ...

PVince81 commented:
Rebase to resurrect Jenkins then squash+merge now and keep the "backport-request" label

DeepDiver1975 commented:
:ghost: 

```
Test\Encryption\Keys\StorageTest::testGetUserKey
OC\User\NoUserException: Backends provided no user object for user1

/var/lib/jenkins/workspace/owncloud-core_core_PR-26820-5ZSS3Y2GKGCIYIEECAW4OPVKEEWW4OIQOCEPDCFQAP5744QLCREA/lib/private/files/filesystem.php:391
/var/lib/jenkins/workspace/owncloud-core_core_PR-26820-5ZSS3Y2GKGCIYIEECAW4OPVKEEWW4OIQOCEPDCFQAP5744QLCREA/lib/private/encryption/keys/storage.php:75
/var/lib/jenkins/workspace/owncloud-core_core_PR-26820-5ZSS3Y2GKGCIYIEECAW4OPVKEEWW4OIQOCEPDCFQAP5744QLCREA/tests/lib/encryption/keys/storage.php:236

DeepDiver1975 committed:
Fix unit test execution

DeepDiver1975 commented on lib/private/encryption/keys/storage.php:
```diff
@@ -70,6 +70,10 @@ public function __construct(View $view, Util $util) {
 	public function getUserKey($uid, $keyId, $encryptionModuleId) {
+		$currentUser = \OC_User::getUser();
+		if (!is_null($uid) && $uid !== '' && $uid !== $currentUser) {
+			\OC\Files\Filesystem::initMountPoints($uid);
```
@PVince81 I guess we should apply this patch to all other methods which call constructUserKeyPath with a user ....

PVince81 replied:
Indeed, makes sense

PVince81 replied:
I looked at the other code path within that class and it seems it would indeed be needed there.
It is not clear whether `initMountPoints` was already called from outside there, it's not a guarantee.
I'll add something in all the methods.

PVince81 commented:
- [ ] add it to other code paths as well

Commit:
Init mount points for user in more places in Keys\Storage

PVince81 commented:
@DeepDiver1975 I've added the init mount points in the other code paths, please check it out: https://github.com/owncloud/core/pull/26820/commits/4b602f654be03b3d7efca7c2bacac38a3a9902ce

A quick manual test with LDAP + homeDirectory shows that it still works.

Now I don't feel comfortable merging this without having automated tests... this would require to finish/solve https://github.com/owncloud/core/pull/26586 and also introduce a new user backend that md5's the user home to simulate this situation. Maybe that backend could be part of the "testing" app ? Or do we make it an official feature configurable in config.php ?

PVince81 commented:
For now I made this PR with the md5 hack to see if the current test coverage would catch the issue: https://github.com/owncloud/core/pull/26844

PVince81 commented:
Let's first get the automated tests sorted out, I want proper coverage for this: https://github.com/owncloud/core/pull/26844

felixboehm commented:
@PVince81 please provide this patch for 9.1

PVince81 commented:
@felixboehm here you go: https://github.com/owncloud/core/pull/26824 (updated with the latest commit).
I retested it with the md5 hack and the fix works.

PVince81 commented:
Looks like we still have some unit test failures

PVince81 commented:
stable9.1: https://github.com/owncloud/core/pull/26824
master: https://github.com/owncloud/core/pull/26917

Automated tests will be added through this PR https://github.com/owncloud/core/pull/26844#issuecomment-271656672.

PVince81 commented:
Jenkins passed.

Since most of the code is now modified by me, need review from someone else please @jvillafanez @DeepDiver1975 @VicDeo @butonic

Commit:
Added test that check if user home is mounted after resolving key

jvillafanez commented:
Same comments as https://github.com/owncloud/core/pull/26917

PVince81 commented:
Cherry-picked the test from master and replace `createMock` with `getMock`.

@jvillafanez please re-review. Thanks

PVince81 commented:
forgot to `git add` ☕️ 

NOW it should be ok

jvillafanez commented:
👍

PVince81 commented:
Let's wait before merging. Need to investigate this (not-so-critical-but-still) regression https://github.com/owncloud/core/issues/26935

Commit:
Don't mount user home with alternate keys root

When encryption keys are stored outside the user's homes, there is no
need to mount said homes.

PVince81 commented:
Yeah, syntax error in PHP 5.4.
Will adjust when rebasing later.

PVince81 commented:
- [ ] TODO: fix unit test syntax error
- [ ] TODO: cherry-pick the regression fix: https://github.com/owncloud/core/pull/26938
- [ ] TODO: cherry-pick the other relevant fix about alternative home dir: https://github.com/owncloud/core/pull/26937

PVince81 commented:
Need rereview, I cherry-picked backports of master to fix a regression and incomplete cases.

Commit:
Ignore exception when deleting keys of deleted user

Whenever a user was deleted for encryption where the keys are stored in
the home, we can ignore user existence exceptions because it means the
keys are already gone.

PVince81 commented:
Rebased and removed the "preDelete" commit and cherry-picked the better fix from https://github.com/owncloud/core/pull/26968

After all these additions, let's retest this:
- [ ] TEST: sharing with alt home works
- [ ] TEST: delete user works with no warnings
- [ ] TEST: delete user works with no warnings when using alternative key storage

PVince81 commented:
- [ ] backport the integration tests from https://github.com/owncloud/core/pull/26844

Commit:
Fix encryption key storage test for older PHP

PVince81 commented:
Integratin tests backport pending to be merged: https://github.com/owncloud/core/pull/26985

jvillafanez commented:
Changes look good 👍
